File: benchmarks/plot_missing_omics_train_test.R
#!/usr/bin/env Rscript
# Generate plots for the missing-omics train/test benchmark.
#
# Inputs:
#   - benchmarks/missing_omics_train_test_results.csv
#
# Outputs:
#   - benchmarks/missing_omics_train_test_plot.png
#
# This is a base-R fallback for environments without Python/matplotlib.

`%||%` <- function(a, b) if (!is.null(a) && nzchar(a)) a else b

# Locate the script directory (works when invoked via Rscript).
full_args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", full_args, value = TRUE)
script_path <- if (length(file_arg) > 0) sub("^--file=", "", file_arg[1]) else ""
here <- if (nzchar(script_path)) dirname(normalizePath(script_path)) else normalizePath("benchmarks", mustWork = FALSE)
results_csv <- file.path(here, "missing_omics_train_test_results.csv")
out_png <- file.path(here, "missing_omics_train_test_plot.png")

if (!file.exists(results_csv)) {
  stop(paste0("Missing input CSV: ", results_csv, "\nRun the benchmark first to generate it."))
}

df <- read.csv(results_csv, stringsAsFactors = FALSE)

num_cols <- c(
  "train_missing_pct",
  "test_missing_pct",
  "ebv_test_total",
  "ebv_test_indirect",
  "epv_test_total",
  "epv_test_trait",
  "time_seconds"
)
for (k in num_cols) {
  if (k %in% names(df)) df[[k]] <- as.numeric(df[[k]])
}

seed <- if ("seed" %in% names(df)) df$seed[1] else NA
chain_length <- if ("chain_length" %in% names(df)) df$chain_length[1] else NA
burnin <- if ("burnin" %in% names(df)) df$burnin[1] else NA
missing_mode <- if ("missing_mode" %in% names(df)) df$missing_mode[1] else ""

groups <- split(df, df$test_missing_pct)
groups <- lapply(groups, function(g) g[order(g$train_missing_pct), , drop = FALSE])

colors <- c(`0` = "#1f77b4", `1` = "#d62728")
labels <- c(`0` = "Test omics 0% missing", `1` = "Test omics 100% missing")

series <- function(test_missing, key) {
  g <- groups[[as.character(test_missing)]]
  if (is.null(g) || nrow(g) == 0) return(list(x = numeric(0), y = numeric(0)))
  xs <- g$train_missing_pct * 100.0
  ys <- g[[key]]
  ys[is.nan(ys)] <- NA
  list(x = xs, y = ys)
}

plot_metric <- function(key, title) {
  plot(
    NA,
    xlim = c(0, 100),
    ylim = c(0, 1),
    xlab = "",
    ylab = "",
    main = title
  )
  grid(col = "gray90")
  for (tm in sort(unique(df$test_missing_pct))) {
    s <- series(tm, key)
    lines(s$x, s$y, type = "o", pch = 16, cex = 0.7, lwd = 2, col = colors[as.character(tm)] %||% "black")
  }
}

png(out_png, width = 1100, height = 800, res = 150)

layout(matrix(c(1, 2, 3, 4, 5, 5), nrow = 3, byrow = TRUE), heights = c(1, 1, 0.8))
par(mar = c(4, 4, 2.5, 1.5), oma = c(0, 0, 3, 0))

plot_metric("ebv_test_total", "EBV (test) vs genetic_total")
legend(
  "bottomleft",
  legend = labels[names(labels) %in% names(groups)],
  col = colors[names(colors) %in% names(groups)],
  lwd = 2,
  pch = 16,
  cex = 0.8,
  bty = "n"
)

plot_metric("ebv_test_indirect", "EBV (test) vs genetic_indirect")
plot_metric("epv_test_total", "EPV (test) vs genetic_total")
plot_metric("epv_test_trait", "EPV (test) vs trait1")

# Runtime plot (spanning both columns)
plot(
  NA,
  xlim = c(0, 100),
  ylim = range(df$time_seconds, finite = TRUE),
  xlab = "Training omics missing (%)",
  ylab = "Seconds",
  main = "Runtime (seconds) per configuration"
)
grid(col = "gray90")
for (tm in sort(unique(df$test_missing_pct))) {
  s <- series(tm, "time_seconds")
  lines(s$x, s$y, type = "o", pch = 16, cex = 0.7, lwd = 2, col = colors[as.character(tm)] %||% "black")
}

mtext(
  sprintf(
    "Missing Omics Train/Test Benchmark (seed=%s, chain=%s, burnin=%s, mode=%s)",
    seed, chain_length, burnin, missing_mode
  ),
  outer = TRUE,
  cex = 1.1
)

dev.off()

cat("Wrote: ", out_png, "\n", sep = "")

File: benchmarks/plot_missing_omics_train_test.py
#!/usr/bin/env python3
"""
Generate plots for the missing-omics train/test benchmark.

Inputs:
  - benchmarks/missing_omics_train_test_results.csv

Outputs:
  - benchmarks/missing_omics_train_test_plot.png
"""

from __future__ import annotations

import csv
import os
import shutil
import tempfile
import atexit
from pathlib import Path


def main() -> None:
    here = Path(__file__).resolve().parent
    results_csv = here / "missing_omics_train_test_results.csv"
    if not results_csv.exists():
        raise SystemExit(
            f"Missing input CSV: {results_csv}\n"
            "Run the benchmark first to generate it."
        )

    # Ensure Matplotlib/fontconfig can write caches within the sandbox.
    mpl_config_dir = Path(tempfile.mkdtemp(prefix="mplconfig-"))
    xdg_cache_home = Path(tempfile.mkdtemp(prefix="xdg-cache-"))
    atexit.register(shutil.rmtree, mpl_config_dir, ignore_errors=True)
    atexit.register(shutil.rmtree, xdg_cache_home, ignore_errors=True)
    os.environ.setdefault("MPLCONFIGDIR", str(mpl_config_dir))
    os.environ.setdefault("XDG_CACHE_HOME", str(xdg_cache_home))

    import matplotlib  # noqa: WPS433

    matplotlib.use("Agg")
    import matplotlib.pyplot as plt  # noqa: WPS433

    rows: list[dict[str, str]] = []
    with results_csv.open(newline="") as f:
        reader = csv.DictReader(f)
        rows = list(reader)

    for r in rows:
        # keys used for plotting
        for k in (
            "train_missing_pct",
            "test_missing_pct",
            "ebv_test_total",
            "ebv_test_indirect",
            "epv_test_total",
            "epv_test_trait",
            "time_seconds",
        ):
            r[k] = float(r[k])  # type: ignore[assignment]

    # Group by test_missing_pct
    groups: dict[float, list[dict[str, float]]] = {}
    for r0 in rows:
        r = {k: float(v) if isinstance(v, (int, float)) else v for k, v in r0.items()}  # type: ignore[arg-type]
        test_missing = float(r["test_missing_pct"])
        groups.setdefault(test_missing, []).append(r)  # type: ignore[arg-type]

    for test_missing, rs in groups.items():
        rs.sort(key=lambda x: float(x["train_missing_pct"]))  # type: ignore[index]
        groups[test_missing] = rs

    # Metadata for title
    seed = int(float(rows[0]["seed"]))  # type: ignore[arg-type]
    chain_length = int(float(rows[0]["chain_length"]))  # type: ignore[arg-type]
    burnin = int(float(rows[0]["burnin"]))  # type: ignore[arg-type]
    missing_mode = str(rows[0].get("missing_mode", ""))

    def series(test_missing: float, key: str) -> tuple[list[float], list[float]]:
        rs = groups.get(test_missing, [])
        xs = [float(r["train_missing_pct"]) * 100.0 for r in rs]  # type: ignore[index]
        ys = [float(r[key]) for r in rs]  # type: ignore[index]
        return xs, ys

    fig = plt.figure(figsize=(11, 8))
    gs = fig.add_gridspec(3, 2, height_ratios=[1, 1, 0.8])
    ax_ebv_total = fig.add_subplot(gs[0, 0])
    ax_ebv_indir = fig.add_subplot(gs[0, 1], sharex=ax_ebv_total)
    ax_epv_total = fig.add_subplot(gs[1, 0], sharex=ax_ebv_total)
    ax_epv_trait = fig.add_subplot(gs[1, 1], sharex=ax_ebv_total)
    ax_time = fig.add_subplot(gs[2, :], sharex=ax_ebv_total)

    fig.suptitle(
        f"Missing Omics Train/Test Benchmark (seed={seed}, chain={chain_length}, burnin={burnin}, mode={missing_mode})"
    )

    colors = {0.0: "#1f77b4", 1.0: "#d62728"}  # blue, red
    labels = {0.0: "Test omics 0% missing", 1.0: "Test omics 100% missing"}

    def plot_metric(ax, key: str, title: str) -> None:
        for test_missing in sorted(groups.keys()):
            xs, ys = series(test_missing, key)
            ax.plot(
                xs,
                ys,
                marker="o",
                linewidth=2,
                markersize=4,
                color=colors.get(test_missing),
                label=labels.get(test_missing, f"test_missing={test_missing}"),
            )
        ax.set_title(title)
        ax.set_ylim(0.0, 1.0)
        ax.grid(True, alpha=0.3)

    plot_metric(ax_ebv_total, "ebv_test_total", "EBV (test) vs genetic_total")
    plot_metric(ax_ebv_indir, "ebv_test_indirect", "EBV (test) vs genetic_indirect")
    plot_metric(ax_epv_total, "epv_test_total", "EPV (test) vs genetic_total")
    plot_metric(ax_epv_trait, "epv_test_trait", "EPV (test) vs trait1")

    # Runtime plot
    for test_missing in sorted(groups.keys()):
        xs, ys = series(test_missing, "time_seconds")
        ax_time.plot(
            xs,
            ys,
            marker="o",
            linewidth=2,
            markersize=4,
            color=colors.get(test_missing),
        )
    ax_time.set_title("Runtime (seconds) per configuration")
    ax_time.set_ylabel("Seconds")
    ax_time.grid(True, alpha=0.3)

    ax_time.set_xlabel("Training omics missing (%)")
    ax_ebv_total.legend(loc="lower left", fontsize=9)

    out_png = here / "missing_omics_train_test_plot.png"
    fig.tight_layout(rect=[0, 0.02, 1, 0.94])
    fig.savefig(out_png, dpi=200)
    print(f"Wrote: {out_png}")


if __name__ == "__main__":
    main()
